File: sources/Android/android-communications/library/src/main/java/com/polar/androidcommunications/api/ble/model/gatt/client/pmd/PmdControlPointResponse.kt
package com.polar.androidcommunications.api.ble.model.gatt.client.pmd

class PmdControlPointResponse(data: ByteArray) {
    companion object {
        const val CONTROL_POINT_RESPONSE_CODE = 0xF0.toByte()
    }

    val responseCode: Byte = data[0]
    val opCode: PmdControlPointCommandClientToService = PmdControlPointCommandClientToService.values()[data[1].toInt()]
    val measurementType: Byte = data[2]
    val status: PmdControlPointResponseCode = PmdControlPointResponseCode.values()[data[3].toInt()]
    val more: Boolean
    var parameters: ByteArray

    enum class PmdControlPointResponseCode(val numVal: Int) {
        SUCCESS(0),
        ERROR_INVALID_OP_CODE(1),
        ERROR_INVALID_MEASUREMENT_TYPE(2),
        ERROR_NOT_SUPPORTED(3),
        ERROR_INVALID_LENGTH(4),
        ERROR_INVALID_PARAMETER(5),
        ERROR_ALREADY_IN_STATE(6),
        ERROR_INVALID_RESOLUTION(7),
        ERROR_INVALID_SAMPLE_RATE(8),
        ERROR_INVALID_RANGE(9),
        ERROR_INVALID_MTU(10),
        ERROR_INVALID_NUMBER_OF_CHANNELS(11),
        ERROR_INVALID_STATE(12),
        ERROR_DEVICE_IN_CHARGER(13),
        ERROR_DISK_FULL(14);
    }

    init {
        var receivedParams = byteArrayOf()
        if (status == PmdControlPointResponseCode.SUCCESS) {
            more = data.size > 4 && data[4] != 0.toByte()
            if (data.size > 5) {
                receivedParams = data.copyOfRange(5, data.size)
            }
        } else {
            more = false
        }
        parameters = receivedParams
    }

    override fun toString(): String {
        return "\n" +
                "responseCode: ${"%02x".format(responseCode)}\n" +
                "opCode: $opCode\n" +
                "measurementType: $measurementType\n" +
                "status: $status\n" +
                "more: $more\n" +
                "parameters: ${parameters.joinToString(separator = " ") { eachByte -> "%02x".format(eachByte) }}"
    }
}

File: sources/iOS/ios-communications/Sources/iOSCommunications/ble/api/model/gatt/client/pmd/PmdControlPointResponse.swift
//  Copyright © 2023 Polar. All rights reserved.

import Foundation

internal struct PmdControlPointResponse {
    static let CONTROL_POINT_RESPONSE_CODE: UInt8 = 0xF0
    public let response: UInt8
    public let opCode: UInt8
    public let type: PmdMeasurementType
    public let errorCode: PmdResponseCode
    public let more: Bool
    public let parameters = NSMutableData()
    public init(_ data: Data) {
        response = data[0]
        opCode = data[1]
        type = PmdMeasurementType(rawValue: data[2]) ?? PmdMeasurementType.unknown_type
        errorCode = PmdResponseCode(rawValue: Int(data[3])) ?? PmdResponseCode.unknown_error
        if data.count > 4 {
            more = data[4] != 0
            parameters.append(data.subdata(in: 5..<data.count))
        } else {
            more = false
        }
    }
}

public enum PmdResponseCode: Int {
    case success = 0
    case errorInvalidOpCode = 1
    case errorInvalidMeasurementType = 2
    case errorNotSupported = 3
    case errorInvalidLength = 4
    case errorInvalidParameter = 5
    case errorAlreadyInState = 6
    case errorInvalidResolution = 7
    case errorInvalidSampleRate = 8
    case errorInvalidRange = 9
    case errorInvalidMTU = 10
    case errorInvalidNumberOfChannels = 11
    case errorInvalidState = 12
    case errorDeviceInCharger = 13
    case errorDiskFull = 14
    case unknown_error = 0xffff
    
    var description : String {
        switch self {
        case .success: return "Success"
        case .errorInvalidOpCode: return "Invalid op code"
        case .errorInvalidMeasurementType: return "Invalid measurement type"
        case .errorNotSupported: return "Not supported"
        case .errorInvalidLength: return "Invalid length"
        case .errorInvalidParameter: return "Invalid parameter"
        case .errorAlreadyInState: return "Already in state"
        case .errorInvalidResolution: return "Invalid Resolution"
        case .errorInvalidSampleRate: return "Invalid Sample rate"
        case .errorInvalidRange: return "Invalid Range"
        case .errorInvalidMTU: return "Invalid MTU"
        case .errorInvalidNumberOfChannels: return "Invalid Number of channels"
        case .errorInvalidState: return "Invalid state"
        case .errorDeviceInCharger: return "Device in charger"
        case .errorDiskFull: return "Disk full"
        case .unknown_error: return "unknown error"
        }
    }
}
